cirq-core/cirq/contrib/paulistring/pauli_string_measurement_with_readout_mitigation.py

```diff
@@ -0,0 +1,312 @@
+
+    for pauli_index, circuit_result in enumerate(circuit_results):
+        measurement_results = circuit_result.measurements["m"]
+
+        mitigated_values, d_m = tensored_cm.readout_mitigation_pauli_uncorrelated(
+            qubits, measurement_results
+        )
+
+        p1 = np.mean(np.sum(measurement_results, axis=1) % 2)
+        unmitigated_values = 1 - 2 * np.mean(p1)
+        d_unmit = 2 * np.sqrt(p1 * (1 - p1) / pauli_repetitions)
+
+        pauli_measurement_results.append(
+            PauliStringMeasurementResult(
+                pauli_string=pauli_strings[pauli_index],
+                mitigated_expectation=mitigated_values,
+                mitigated_stddev=d_m,
+                unmitigated_expectation=unmitigated_values,
+                unmitigated_stddev=d_unmit,
+            )
+        )
+
+    return pauli_measurement_results
+
+
+def measure_pauli_strings(
+    circuits_to_pauli: Dict[circuits.FrozenCircuit, list[ops.PauliString]],
+    sampler: work.Sampler,
+    rng_or_seed: Union[np.random.Generator, int],
+    pauli_repetitions: int,
+    readout_repetitions: int,
+    num_random_bitstrings: int,
+) -> List[CircuitToPauliStringsMeasurementResult]:
+    """Measures expectation values of Pauli strings on given circuits with/without
+    readout error mitigation.
+
+    This function takes a list of circuits and corresponding List[Pauli string] to measure.
+    For each circuit-List[Pauli string] pair, it:
+    1.  Constructs circuits to measure the Pauli string expectation value by
+        adding basis change moments and measurement operations.
+    2.  Runs shuffled readout benchmarking on these circuits to calibrate readout errors.
+    3.  Mitigates readout errors using the calibrated confusion matrices.
+    4.  Calculates and returns both error-mitigated and unmitigatedexpectation values for
+    each Pauli string.
+
+    Args:
+        circuits_to_pauli: A dictionary mapping circuits to a list of Pauli strings
+            to measure.
+        sampler: The sampler to use.
+        rng_or_seed: A random number generator or seed for the shuffled benchmarking.
+        pauli_repetitions: The number of repetitions for each circuit when measuring
+            Pauli strings.
+        readout_repetitions: The number of repetitions for readout calibration
+            in the shuffled benchmarking.
+        num_random_bitstrings: The number of random bitstrings to use in shuffled
+            benchmarking.
+
+    Returns:
+        A list of CircuitToPauliStringsMeasurementResult objects, where each object contains:
+            - The circuit that was measured.
+            - A list of PauliStringMeasurementResult objects.
+            - The calibration result for single-qubit readout errors.
+    """
+
+    # Extract unique qubit tuples from all circuits
+    unique_qubit_tuples = set()
+    for circuit in circuits_to_pauli.keys():
+        unique_qubit_tuples.add(tuple(sorted(set(circuit.all_qubits()))))
```

Comment:
`circuit.all_qubits()` returns a frozenset, so you can dop the `set` here and elsewhere